keras_preprocessing/image/utils.py

```diff
@@ -109,32 +110,33 @@ def load_img(path, grayscale=False, color_mode='rgb', target_size=None,
     if pil_image is None:
         raise ImportError('Could not import PIL.Image. '
                           'The use of `load_img` requires PIL.')
-    img = pil_image.open(path)
```

Comment:
you can probably do `with pil_image.close(path) as img`, keeping everything in PILLOW land. Can you check?

Reply:
@rragundez Testing with `with pil_image.open(path) as img:`, I get the following errors:

<details>
 <summary>Error Report</summary>
(3.5) Duhaime:keras-preprocessing doug$ py.test tests/
================================================== test session starts ==================================================
platform darwin -- Python 3.5.3, pytest-3.0.6, py-1.8.0, pluggy-0.4.0
rootdir: /Users/doug/Desktop/keras-preprocessing, inifile:
plugins: pep8-1.0.6, cov-2.6.0, celery-4.1.1
collected 79 items

tests/sequence_test.py .........
tests/test_api.py ...
tests/test_documentation.py .
tests/text_test.py ....................
tests/image/affine_transformations_test.py ........
tests/image/dataframe_iterator_test.py ..............
tests/image/directory_iterator_test.py .....
tests/image/image_data_generator_test.py ..........
tests/image/iterator_test.py .
tests/image/numpy_array_iterator_test.py .
tests/image/test_image_api.py ..
tests/image/utils_test.py .F...

======================================================= FAILURES ========================================================
_____________________________________________________ test_load_img _____________________________________________________

tmpdir = local('/private/var/folders/cy/ywx04ttn12n_xmgjf5s6sbzr0000gn/T/pytest-of-doug/pytest-49/test_load_img0')

    def test_load_img(tmpdir):
        filename_rgb = str(tmpdir / 'rgb_utils.png')
        filename_rgba = str(tmpdir / 'rgba_utils.png')
        filename_grayscale_8bit = str(tmpdir / 'grayscale_8bit_utils.png')
        filename_grayscale_16bit = str(tmpdir / 'grayscale_16bit_utils.tiff')
        filename_grayscale_32bit = str(tmpdir / 'grayscale_32bit_utils.tiff')

        original_rgb_array = np.array(255 * np.random.rand(100, 100, 3),
                                      dtype=np.uint8)
        original_rgb = utils.array_to_img(original_rgb_array, scale=False)
        original_rgb.save(filename_rgb)

        original_rgba_array = np.array(255 * np.random.rand(100, 100, 4),
                                       dtype=np.uint8)
        original_rgba = utils.array_to_img(original_rgba_array, scale=False)
        original_rgba.save(filename_rgba)

        original_grayscale_8bit_array = np.array(255 * np.random.rand(100, 100, 1),
                                                 dtype=np.uint8)
        original_grayscale_8bit = utils.array_to_img(original_grayscale_8bit_array,
                                                     scale=False)
        original_grayscale_8bit.save(filename_grayscale_8bit)

        original_grayscale_16bit_array = np.array(
            np.random.randint(-2147483648, 2147483647, (100, 100, 1)), dtype=np.int16
        )
        original_grayscale_16bit = utils.array_to_img(original_grayscale_16bit_array,
                                                      scale=False, dtype='int16')
        original_grayscale_16bit.save(filename_grayscale_16bit)

        original_grayscale_32bit_array = np.array(
            np.random.randint(-2147483648, 2147483647, (100, 100, 1)), dtype=np.int32
        )
        original_grayscale_32bit = utils.array_to_img(original_grayscale_32bit_array,
                                                      scale=False, dtype='int32')
        original_grayscale_32bit.save(filename_grayscale_32bit)

        # Test that loaded image is exactly equal to original.

        loaded_im = utils.load_img(filename_rgb)
>       loaded_im_array = utils.img_to_array(loaded_im)

tests/image/utils_test.py:61:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/Users/doug/anaconda/envs/3.5/lib/python3.5/site-packages/Keras_Preprocessing-1.1.0-py3.5.egg/keras_preprocessing/image/utils.py:309: in img_to_array
    x = np.asarray(img, dtype=dtype)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

a = <PIL.PngImagePlugin.PngImageFile image mode=RGB size=100x100 at 0x127774208>, dtype = 'float32', order = None

    @set_module('numpy')
    def asarray(a, dtype=None, order=None):
        """Convert the input to an array.

        Parameters
        ----------
        a : array_like
            Input data, in any form that can be converted to an array.  This
            includes lists, lists of tuples, tuples, tuples of tuples, tuples
            of lists and ndarrays.
        dtype : data-type, optional
            By default, the data-type is inferred from the input data.
        order : {'C', 'F'}, optional
            Whether to use row-major (C-style) or
            column-major (Fortran-style) memory representation.
            Defaults to 'C'.

        Returns
        -------
        out : ndarray
            Array interpretation of `a`.  No copy is performed if the input
            is already an ndarray with matching dtype and order.  If `a` is a
            subclass of ndarray, a base class ndarray is returned.

        See Also
        --------
        asanyarray : Similar function which passes through subclasses.
        ascontiguousarray : Convert input to a contiguous array.
        asfarray : Convert input to a floating point ndarray.
        asfortranarray : Convert input to an ndarray with column-major
                         memory order.
        asarray_chkfinite : Similar function which checks input for NaNs and Infs.
        fromiter : Create an array from an iterator.
        fromfunction : Construct an array by executing a function on grid
                       positions.

        Examples
        --------
        Convert a list into an array:

        >>> a = [1, 2]
        >>> np.asarray(a)
        array([1, 2])

        Existing arrays are not copied:

        >>> a = np.array([1, 2])
        >>> np.asarray(a) is a
        True

        If `dtype` is set, array is copied only if dtype does not match:

        >>> a = np.array([1, 2], dtype=np.float32)
        >>> np.asarray(a, dtype=np.float32) is a
        True
        >>> np.asarray(a, dtype=np.float64) is a
        False

        Contrary to `asanyarray`, ndarray subclasses are not passed through:

        >>> issubclass(np.recarray, np.ndarray)
        True
        >>> a = np.array([(1.0, 2), (3.0, 4)], dtype='f4,i4').view(np.recarray)
        >>> np.asarray(a) is a
        False
        >>> np.asanyarray(a) is a
        True

        """
>       return array(a, dtype, copy=False, order=order)
E       TypeError: float() argument must be a string or a number, not 'PngImageFile'

../../anaconda/envs/3.5/lib/python3.5/site-packages/numpy/core/numeric.py:538: TypeError
========================================= 1 failed, 78 passed in 12.10 seconds ==========================================
</details>

I'm happy to try other potential solutions, but just wanted to mention that the method in this pull request is the [only self-closing method endorsed by the Pillow docs](https://pillow.readthedocs.io/en/4.0.x/reference/open_files.html).